### PR TITLE
[DPE-1557] - HA tests - network cut - with IP change

### PR DESCRIPTION
**Pull request by Mehdi-Bendriss**
## Issue
This PR implements the HA test and fixes the bugs raised with it for [DPE-1557](https://warthogs.atlassian.net/browse/DPE-1557), namely - this PR implements:
- Integration tests for:
    - network cut - with IP change - on node with elected CM
    - network cut - with IP change - on node with primary shard of index
    - helpers for accurately retrieving all units related information
- implementation:
    - renewal of certificates for node after the IP change 
    - renewal of the `network.host` content with the new hosts
    - manual trigger of topology re-computing after network restored
- fixes:
    - duplication issue of the content of the `jvm.options` file 
    - infinite loop of the opensearch daemon restart
    - tls cert key format
    - only mark a unit as `ready` when it's fully able to take on requests 
    - unit tests

[DPE-1557]: https://warthogs.atlassian.net/browse/DPE-1557?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib/charms/opensearch/v0/opensearch_tls.py [hunk 1]
```diff
@@ -85,6 +85,18 @@ def _on_set_tls_private_key(self, event: ActionEvent) -> None:
...
             event.fail(str(e))
 
+    def request_new_unit_certificates(self) -> None:
+        """Requests a new certificate with the given scope and type from the tls operator."""
+        for cert_type in [CertType.UNIT_HTTP, CertType.UNIT_TRANSPORT]:
+            csr = self.charm.secrets.get_object(Scope.UNIT, cert_type.val)["csr"].encode("utf-8")
+            self.certs.request_certificate_revocation(csr)
+
+        # doing this sequentially (revoking -> requesting new ones), to avoid triggering
+        # the "certificate available" callback with old certificates
```
Comment (WRFitch) [inline lib/charms/opensearch/v0/opensearch_tls.py:95]: nice

File: tests/unit/lib/test_opensearch_base_charm.py [hunk 2]
```diff
@@ -141,7 +143,8 @@ def test_on_start(
             self.peers_data.delete(Scope.APP, "security_index_initialised")
             _can_service_start.return_value = True
             self.harness.set_leader(True)
-            start.reset_mock()
+            # start.reset_mock()
```
Comment (WRFitch) [inline tests/unit/lib/test_opensearch_base_charm.py:146]: Is this still necessary?

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 3]
```diff
@@ -536,21 +558,29 @@ def _start_opensearch(self, event: EventBase) -> None:  # noqa: C901
...
+        # initialize the security index if needed (and certs written on disk etc.)
+        if self.unit.is_leader() and not self.peers_data.get(
+            Scope.APP, "security_index_initialised"
+        ):
+            admin_secrets = self.secrets.get_object(Scope.APP, CertType.APP_ADMIN.val)
+            self._initialize_security_index(admin_secrets)
+            self.peers_data.put(Scope.APP, "security_index_initialised", True)
+
+        # it sometimes takes a few seconds before the node is fully "up" otherwise a 503 error
+        # may be thrown when calling a node - we want to unsure this node is perfectly ready
```
Comment (WRFitch) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:580]:
```suggestion
        # may be thrown when calling a node - we want to ensure this node is perfectly ready
```

**Review by WRFitch — APPROVED** [review]
great stuff Mehdi, a couple of minor comments:

File: tests/integration/helpers.py [hunk 4]
```diff
@@ -186,21 +229,36 @@ def get_application_unit_ids_ips(ops_test: OpsTest, app: str = APP_NAME) -> Dict
...
+
+async def get_application_unit_ids_hostnames(
+    ops_test: OpsTest, app: str = APP_NAME
+) -> Dict[int, str]:
+    """List the units of an application by id and corresponding host name."""
+    result = {}
     for unit in ops_test.model.applications[app].units:
-        result[int(unit.name.split("/")[1])] = unit.public_address
+        unit_id = int(unit.name.split("/")[1])
+        _, raw_hostname, _ = await ops_test.juju("ssh", f"{app}/{unit_id}", "hostname")
```
Comment (delgod) [inline tests/integration/helpers.py:245]: I don't see where we use `raw_hostname`. I think this line is replaced by the next line

**Review by delgod — APPROVED** [review]

**Pull request merged**